larq_compute_engine/tflite/tests/bconv2d_test.cc

```diff
@@ -383,6 +408,134 @@ MATCHER_P(FloatNearPointwise, tol, "Out of range") {
           std::get<0>(arg) < std::get<1>(arg) + tol);
 }
 
+template <typename TInput, typename TOutput>
+inline void set_lce_op_input(const RuntimeShape& input_shape,
+                             std::vector<float> input_data,
+                             BConv2DOpModel<TInput, TOutput>& m_lce) {
+  std::vector<TInput> input_data_bp;
+  std::size_t input_rows_bp, input_cols_bp, input_bitpadding;
+  core::packbits_matrix<core::BitpackOrder::Canonical>(
+      input_data.data(), FlatSizeSkipDim(input_shape, 3), input_shape.Dims(3),
+      input_data_bp, input_rows_bp, input_cols_bp, input_bitpadding,
+      core::Axis::RowWise);
+  m_lce.SetInput(input_data_bp);
+}
+
+// A specialisation for float input  (i.e., not reading bitpacked input).
+template <typename TOutput>
+inline void set_lce_op_input(const RuntimeShape& input_shape,
+                             std::vector<float> input_data,
+                             BConv2DOpModel<float, TOutput>& m_lce) {
+  m_lce.SetInput(input_data);
+}
+
+template <typename TOutput>
+inline void test_lce_op_output(const std::vector<TOutput>& lce_output_data,
+                               const std::vector<int>& builtin_output_shape,
+                               const std::vector<float>& builtin_output_data) {
+  // Apply bitpacking to the builtin output.
+  std::vector<TOutput> builtin_output_data_bp;
+  std::size_t output_rows_bp, output_cols_bp, output_bitpadding;
+  core::packbits_matrix<core::BitpackOrder::Canonical>(
+      builtin_output_data.data(),
+      builtin_output_shape.at(0) * builtin_output_shape.at(1) *
+          builtin_output_shape.at(2),
+      builtin_output_shape.at(3), builtin_output_data_bp, output_rows_bp,
+      output_cols_bp, output_bitpadding, core::Axis::RowWise);
+
+  // We need the outputs here to be bit-exact, so don't allow for floating
+  // point imprecision.
```

Comment:
I'm wondering if this would fail if the float in the builtin op was `-0.00001f` and the float in our op was `0.00001f`. I think its fine for now, but if this happens in the future we would have to add an extra loop here that manually checks the bits and allows for a mismatch in that case.

Reply:
Ah that is a good point. I guess the likelihood of that happening is quite slim but iterating though and checking individually would not be hard.